# Fundamentals: Python Function

## Warmup: adder function

Create a function that takes two numbers as arguments and return their sum.

### Examples

- `addition(3, 2) ➞ 5`
- `addition(-3, -6) ➞ -9`
- `addition(7, 3) ➞ 10`

In [5]:
# write your function here

def f(a,b):
    localList = [a,b]
    return sum(localList)

print(f(1,2))

3


In [7]:
# Warmup 2: even numbers
# Generate a Python list of all the even numbers between 4 to 3048

evenNumbGenList = [i for i in range(4, 3049) if i%2==0]
print(evenNumbGenList)

[4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42, 44, 46, 48, 50, 52, 54, 56, 58, 60, 62, 64, 66, 68, 70, 72, 74, 76, 78, 80, 82, 84, 86, 88, 90, 92, 94, 96, 98, 100, 102, 104, 106, 108, 110, 112, 114, 116, 118, 120, 122, 124, 126, 128, 130, 132, 134, 136, 138, 140, 142, 144, 146, 148, 150, 152, 154, 156, 158, 160, 162, 164, 166, 168, 170, 172, 174, 176, 178, 180, 182, 184, 186, 188, 190, 192, 194, 196, 198, 200, 202, 204, 206, 208, 210, 212, 214, 216, 218, 220, 222, 224, 226, 228, 230, 232, 234, 236, 238, 240, 242, 244, 246, 248, 250, 252, 254, 256, 258, 260, 262, 264, 266, 268, 270, 272, 274, 276, 278, 280, 282, 284, 286, 288, 290, 292, 294, 296, 298, 300, 302, 304, 306, 308, 310, 312, 314, 316, 318, 320, 322, 324, 326, 328, 330, 332, 334, 336, 338, 340, 342, 344, 346, 348, 350, 352, 354, 356, 358, 360, 362, 364, 366, 368, 370, 372, 374, 376, 378, 380, 382, 384, 386, 388, 390, 392, 394, 396, 398, 400, 402, 404, 406, 408, 410, 412, 414, 416, 418, 420, 422, 

## Exercise 1: First and Last 

Creates a function that takes a string and returns the first and last character.
### Examples

```
first_last("ganesh") ➞ "gh"

first_last("kali") ➞ "ki"

first_last("shiva") ➞ "sa"

first_last("vishnu") ➞ "vu"

first_last("durga") ➞ "da"
```

In [11]:
# exercise 1
def firstLastStringChar(s):
    first = s[0]
    last = s[-1]
    return first+last
print(firstLastStringChar('roger'))



rr
<class 'str'>


## Exercise 2: Keys in Dict

Write a function that returns `True` if a dictionary contains the specified key, and `False` otherwise.

### Examples

```
has_key({ "a": 44, "b": 45, "c": 46 }, "d") ➞ False

has_key({ "craves": True, "midnight": True, "snack": True }, "morning") ➞ False

has_key({ "pot": 1, "tot": 2, "not": 3 }, "not") ➞ True
```

In [26]:
# exercise 2
def has_key(myDic, myKey):
    ret = None
    for k in myDic.keys():
        if k == myKey:
            ret = True
            break
    else:
        ret = False
    return ret

dic1 = { "a": 44, "b": 45, "c": 46 }
dic2 = { "craves": True, "midnight": True, "snack": True }
dic3 = { "pot": 1, "tot": 2, "not": 3 }


print(has_key(dic1, "d"))
print(has_key(dic2, "morning"))
print(has_key(dic3, "not"))

False
False
True


## Exercise 3: Test Distinct

Write a Python function that takes a sequence of numbers and determines whether all the numbers are different from each other.

```
test_distinct([2,3,4,5,7,8]) -> True
test_distinct([2,3,4,5,8,8]) -> False
```
def test_distinct(myList):
    ret = None
    size_myList = len(myList)
    rem_duplicates = len(set(myList))
    if size_myList == rem_duplicates:
        ret = "Unique numbers only"
    else:
        ret = "Contains duplicates"
    return ret
    
test_distinct([2,3,4,5,7,8])
test_distinct([2,3,4,5,8,8])

In [33]:
# exercise 3
def test_distinct(myList): 
    ret = None 
    size_myList = len(myList) 
    size_rem_duplicates = len(set(myList)) 
    
    if size_myList == size_rem_duplicates: 
        ret = "Unique numbers only" 
    else: 
        ret = "Contains duplicates" 
    return ret

#test_distinct([2,3,4,5,7,8]) 
test_distinct([2,3,4,5,8,8])

'Contains duplicates'

## Exercise 4: Spell Checker

Create a function that takes in two lists: 

1. A sentence as a list of user-typed words

2. The list of correctly-typed words

and outputs a list containing 1s (correctly-typed words) and 0s (incorrectly-typed words).

### Inputs:
> User-typed: \["cat", "blue", "skt", "umbrells", "paddy"\]

> Correct: \["cat", "blue", "sky", "umbrella", "paddy"\]

### Output: \[ 1, 1, 0, 0, 1 \]

### Examples

```
correct_stream(
  ["it", "is", "find"],
  ["it", "is", "fine"]
) ➞ [1, 1, 0]

correct_stream(
  ["april", "showrs", "bring", "may", "flowers"],
  ["april", "showers", "bring", "may", "flowers"]
) ➞ [1, 0, 1, 1, 1]

```

In [49]:
# exercise 4
import numpy as np

def spellChecker ( userList, correctList):
    userList = np.array(userList)
    correctList = np.array(correctList)
    test = userList == correctList
    ret = []
    for i in range(0,len(test)):
        if test[i] == True:
            ret.append(1)
        else:
            ret.append(0)
            
    return ret

#spellChecker(["it", "is", "find"],["it", "is", "fine"])
spellChecker(["april", "showrs", "bring", "may", "flowers"],["april", "showers", "bring", "may", "flowers"])
    

[1, 0, 1, 1, 1]

## Exercise 5: FizzBuzz

Fizzbuzz is a game for children and a drinking game for adults. It's also a classic programming interview question.

Suppose we have a number n. We have to display a string representation of all numbers from 1 to n, but there are some constraints.

    If the number is divisible by 3, write Fizz instead of the number
    If the number is divisible by 5, write Buzz instead of the number
    If the number is divisible by 3 and 5 both, write FizzBuzz instead of the number

So the results are `1, 2, Fizz, 4, Buzz, ..., 14, Fizzbuzz, 16, 17, Fizz, 19, Buzz, ...`

Write a python program that prints fizzbuzz for numbers 1 to 150

In [50]:
# exercise 5
for i in range(1, 151):
    ret = i
    if i%3==0 and i%5==0:
        ret='FizzBuzz'
    elif i%3==0:
        ret='Fizz'
    elif i%5==0:
        ret='Buzz'
    else:
        ret=i
        
    print(ret)


1
2
Fizz
4
Buzz
Fizz
7
8
Fizz
Buzz
11
Fizz
13
14
FizzBuzz
16
17
Fizz
19
Buzz
Fizz
22
23
Fizz
Buzz
26
Fizz
28
29
FizzBuzz
31
32
Fizz
34
Buzz
Fizz
37
38
Fizz
Buzz
41
Fizz
43
44
FizzBuzz
46
47
Fizz
49
Buzz
Fizz
52
53
Fizz
Buzz
56
Fizz
58
59
FizzBuzz
61
62
Fizz
64
Buzz
Fizz
67
68
Fizz
Buzz
71
Fizz
73
74
FizzBuzz
76
77
Fizz
79
Buzz
Fizz
82
83
Fizz
Buzz
86
Fizz
88
89
FizzBuzz
91
92
Fizz
94
Buzz
Fizz
97
98
Fizz
Buzz
101
Fizz
103
104
FizzBuzz
106
107
Fizz
109
Buzz
Fizz
112
113
Fizz
Buzz
116
Fizz
118
119
FizzBuzz
121
122
Fizz
124
Buzz
Fizz
127
128
Fizz
Buzz
131
Fizz
133
134
FizzBuzz
136
137
Fizz
139
Buzz
Fizz
142
143
Fizz
Buzz
146
Fizz
148
149
FizzBuzz


## Exercise 6: Palindrome?

A palindrome is a word which reads the same backward as forwards. Write a function that returns True if a string is a palindrome:

    “laval” ⇒ Yes
    “montreal” ⇒ No

In [76]:
# exercise 6
import numpy as np

#created a Function to reverse a string
def revertString(myString):
    ret=[myString[i] for i in range(len(myString)-1,-1,-1)]
    return ret

#myString = 'laval'
myString = 'montreal'

np_original = np.array(list(myString))
np_reverted = np.array(list(revertString(myString)))
#test will contain a numpy list containing True or False whenever the 2 characters at a specific index are equal
test = np_original == np_reverted

# we want to print Yes or No rather than a list of True or False
if False in test:
    ret = 'No'
else:
    ret = 'Yes'
print(ret)

   
    

No


## Exercise 7: Counting Country IDs

We have a list of user IDS with identifiers of form "userid-COUNTRYID". Calculate how many countries we have.

Ex:

`[gyhwefy7832ryyg-US, iu7ytf87gyuwef-CA, 98yfyg82t3r-US]` -> `{'US': 2, 'CA': 1}`

```
[gyhwefy7832ryyg-US, iu7ytf87gyuwef-CA, 98yfyg82t3r-US,
    iuhsdf897y3righsfe-US, iushdf98h23ryg-IT, GHsg1s54sda8-CA, iuhdf9h32ryu9gf-IT,
    IUufgh98y34-FR, 98yfdiuhdfbviuh8-UK, 45asdjJHHghsdas-CA
] -> {'US': 3, 'CA': 3, "IT": 2, 'UK':1, 'FR':1}
```

In [112]:
# exercise 7

def buildListOfCountries(myList):
    listOfCountries=[]
    
    for elt in myList:
        
        #1-find the hyphen in an element
        hyphPosition = elt.index('-')
        #2-read the country ID by slicing
        countryID = elt[hyphPosition+1:]
        #3-add the country to the local list
        listOfCountries.append(countryID)
    return listOfCountries


def addUpCountry(listCountries):
    dic={}
    for country in listCountries:
        if country in dic.keys():
            dic[country]+=1
        else:
            dic.update({country:1})
    return dic


#useridCOUNTRYID = ['gyhwefy7832ryyg-US', 'iu7ytf87gyuwef-CA', '98yfyg82t3r-US']
useridCOUNTRYID = ['gyhwefy7832ryyg-US', 'iu7ytf87gyuwef-CA', '98yfyg82t3r-US','iuhsdf897y3righsfe-US', 'iushdf98h23ryg-IT', 'GHsg1s54sda8-CA', 'iuhdf9h32ryu9gf-IT', 'IUufgh98y34-FR', '98yfdiuhdfbviuh8-UK', '45asdjJHHghsdas-CA']

listCountries = buildListOfCountries(useridCOUNTRYID)
print(addUpCountry(listCountries))




{'US': 3, 'CA': 3, 'IT': 2, 'FR': 1, 'UK': 1}


## Exercise 8: Docstrings

Write the completed docstrings for each function covered in the "Big O Notation Examples" Section from lecture.

Docstring format is as follows:

```
"""[Summary]

:param [ParamName]: [ParamDescription], defaults to [DefaultParamVal]
:type [ParamName]: [ParamType](, optional)
...
:raises [ErrorType]: [ErrorDescription]
...
:return: [ReturnDescription]
:rtype: [ReturnType]
"""
```

In [107]:
def multiply(val_1, val_2):
    """
    [Summary]
    :param [val_1]:[first value]
    :type [val_1] : [float]
    :param [val_2]:[second value]
    :type [val_2] : [float]
    ...
    :raises [TypeError]:[wrong type for the attribute]
    ... 
    :return: [the mathematical multiplication of val_1 by val_2]
    :rtype: [float]
    
    """
    return val_1 * val_2

help(multiply)



TypeError: multiply() missing 1 required positional argument: 'val_2'

In [109]:
def power(val, exponent):
    """
    [Summary]
    :param [val]:[ value]
    :type [val] : [float]
    :param [exponent]:[exponent value]
    :type [exponent] : [integer]
    ...
    :raises [TypeError]:[wrong type for the attribute]
    ... 
    :return: [return the val raised to the power of exponent]
    :rtype: [float]
    
    """
    
    return val**exponent

#help(power)


4.0

In [110]:
def last_value_in_list(lst):
    """
    [Summary]
    :param [lst]:[ list parameter]
    :type [lst] : [list type]
    
    ...
    :raises [TypeError]:[wrong type for the attribute]
    ... 
    :return: [return the value in the list]
    :rtype: [type of the last element in the list]
    
    """
    
    
    return lst[-1]

help(last_value_in_list)

Help on function last_value_in_list in module __main__:

last_value_in_list(lst)
    [Summary]
    :param [lst]:[ list parameter]
    :type [lst] : [list type]
    
    ...
    :raises [TypeError]:[wrong type for the attribute]
    ... 
    :return: [return the value in the list]
    :rtype: [type of the last element in the list]



In [111]:
def reverse_vals(lst):
    """
    [Summary]
    :param [lst]:[ list parameter]
    :type [lst] : [list type]
    
    ...
    :raises [TypeError]:[wrong type for the attribute]
    ... 
    :return: [return the reversed list]
    :rtype: [list type]
    
    """
    
    return lst[::-1]

help(reverse_vals)

Help on function reverse_vals in module __main__:

reverse_vals(lst)
    [Summary]
    :param [lst]:[ list parameter]
    :type [lst] : [list type]
    
    ...
    :raises [TypeError]:[wrong type for the attribute]
    ... 
    :return: [return the reversed list]
    :rtype: [list type]

